codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/waiters/RustWaiterMatcherGenerator.kt

```diff
@@ -0,0 +1,190 @@
+import software.amazon.smithy.codegen.core.Symbol
+import software.amazon.smithy.jmespath.JmespathExpression
+import software.amazon.smithy.model.node.Node
+import software.amazon.smithy.model.shapes.Shape
+import software.amazon.smithy.rust.codegen.client.smithy.ClientCodegenContext
+import software.amazon.smithy.rust.codegen.client.smithy.ClientRustModule
+import software.amazon.smithy.rust.codegen.core.rustlang.RustModule
+import software.amazon.smithy.rust.codegen.core.rustlang.RustType
+import software.amazon.smithy.rust.codegen.core.rustlang.RustWriter
+import software.amazon.smithy.rust.codegen.core.rustlang.replaceLifetimes
+import software.amazon.smithy.rust.codegen.core.rustlang.rust
+import software.amazon.smithy.rust.codegen.core.rustlang.rustBlockTemplate
+import software.amazon.smithy.rust.codegen.core.rustlang.rustTemplate
+import software.amazon.smithy.rust.codegen.core.rustlang.writable
+import software.amazon.smithy.rust.codegen.core.smithy.RuntimeType
+import software.amazon.smithy.rust.codegen.core.smithy.RuntimeType.Companion.preludeScope
+import software.amazon.smithy.rust.codegen.core.util.dq
+import software.amazon.smithy.rust.codegen.core.util.toSnakeCase
+import software.amazon.smithy.waiters.Matcher
+import software.amazon.smithy.waiters.Matcher.ErrorTypeMember
+import software.amazon.smithy.waiters.Matcher.InputOutputMember
+import software.amazon.smithy.waiters.Matcher.OutputMember
+import software.amazon.smithy.waiters.Matcher.SuccessMember
+import software.amazon.smithy.waiters.PathComparator
+import java.security.MessageDigest
+
+private typealias Scope = Array<Pair<String, Any>>
+
+/**
+ * Generates the Rust code for the Smithy waiter "matcher union".
+ * See https://smithy.io/1.0/spec/waiters.html#matcher-union
+ */
+class RustWaiterMatcherGenerator(
+    private val codegenContext: ClientCodegenContext,
+    private val operationName: String,
+    private val inputShape: Shape,
+    private val outputShape: Shape,
+) {
+    private val runtimeConfig = codegenContext.runtimeConfig
+    private val module = RustModule.pubCrate("matchers", ClientRustModule.waiters)
+    private val inputSymbol = codegenContext.symbolProvider.toSymbol(inputShape)
+    private val outputSymbol = codegenContext.symbolProvider.toSymbol(outputShape)
+
+    fun generate(
+        errorSymbol: Symbol,
+        matcher: Matcher<*>,
+    ): RuntimeType {
+        val fnName = fnName(operationName, matcher)
+        val scope =
+            arrayOf(
+                *preludeScope,
+                "Input" to inputSymbol,
+                "Output" to outputSymbol,
+                "Error" to errorSymbol,
+                "ProvideErrorMetadata" to RuntimeType.provideErrorMetadataTrait(runtimeConfig),
+            )
+        return RuntimeType.forInlineFun(fnName, module) {
+            rustBlockTemplate("pub(crate) fn $fnName(_input: &#{Input}, _result: &#{Result}<#{Output}, #{Error}>) -> bool", *scope) {
```

Comment:
**question**: So we're generating functions like `pub(crate) fn match_test_operation_1555d9862d8934ff9(...)`, can we generate any kind of provenance doc comment that ties it to the model (to aid in debugging/traceability)? 


In Kotlin these matcher functions are generated inline where they are used so they don't need an identifier. I think it's fine to generate them this way but I am wondering more generally what is the design for how these matches will be consumed? What will a generated waiter look like? How will these matchers be tied to an acceptor and retry strategy?

Reply:
I'll add a comment above the function with the matcher json, at the very least. I imagine it will be pretty easy to identify them if you're going from the direction of waiter impl -> matcher. The acceptor/retry strategy will be calling these functions directly.